### PR TITLE
Step 1.5

### PR DESCRIPTION
**Pull request by Barsik-sus**

### CODE REVIEW

File: 1_concepts/1_5_convert_cast_deref/src/main.rs [hunk 1]
```diff
@@ -1,4 +1,121 @@
...
+      Err( InvalidEmail ),
+      EmailString::try_from( "some.email_address5@@gmail.com" )
+    );
+  }
+
+  // this test works for seed = 92
+  #[ test ]
+  fn random_gets_all_values()
+  {
+    let rand = Random::new( 1, 2, 3 );
```
Comment (dmvict) [inline 1_concepts/1_5_convert_cast_deref/src/main.rs:108]: The own implementation of random numbers is good as practice but it is bad as a real code.

File: 1_concepts/1_5_convert_cast_deref/src/main.rs [hunk 2]
```diff
@@ -1,4 +1,121 @@
...
+      )
+    }
+  }
+}
+
+impl< T > Deref for Random< T >
+{
+  type Target = T;
+
+  fn deref( &self ) -> &Self::Target {
```
Comment (dmvict) [inline 1_concepts/1_5_convert_cast_deref/src/main.rs]: Please, update style.

File: 1_concepts/1_5_convert_cast_deref/src/main.rs [hunk 3]
```diff
@@ -1,4 +1,120 @@
...
+      Err( InvalidEmail )
+    }
+  }
 }
+
+#[ derive( Debug ) ]
+pub struct Random< T >
+{
+  pub values : [ T; 3 ],
+  _rng : RefCell< Box< ThreadRng > >
```
Comment (dmvict) [inline 1_concepts/1_5_convert_cast_deref/src/main.rs:31]: Please, explain the reasons of using wrappers `RefCell` and `Box`.

File: 1_concepts/1_5_convert_cast_deref/src/main.rs [hunk 4]
```diff
@@ -1,4 +1,120 @@
...
+
+impl< T > Deref for Random< T >
+{
+  type Target = T;
+
+  fn deref( &self ) -> &Self::Target 
+  {
+    &self.values.choose
+    (
+      &mut self._rng.borrow_mut().as_mut()
```
Comment (dmvict) [inline 1_concepts/1_5_convert_cast_deref/src/main.rs:54]: I see no upper bound of range. 

Can you explain me how it is restricted?

Comment (Barsik-sus) [reply 1_concepts/1_5_convert_cast_deref/src/main.rs:54]: I used "choose" method which returns a reference to random one element

**Comment by dmvict** [comment]
Can be merged.

**Pull request merged**